src/WinRT.Runtime/IWinRTObject.net5.cs

```diff
@@ -1,217 +1,298 @@
                 IObjectReference typedObjRef = (IObjectReference)typeof(IObjectReference).GetMethod("As", Type.EmptyTypes).MakeGenericMethod(vftblType).Invoke(objRef, null);
-#pragma warning restore IL3050
-                if (!QueryInterfaceCache.TryAdd(interfaceType, typedObjRef))
-                {
-                    typedObjRef.Dispose();
-                }
-                return true;
-            }
-        }
-
-        RuntimeTypeHandle IDynamicInterfaceCastable.GetInterfaceImplementation(RuntimeTypeHandle interfaceType)
-        {
-            var type = Type.GetTypeFromHandle(interfaceType);
-            var helperType = type.GetHelperType();
-            if (helperType.IsInterface)
-                return helperType.TypeHandle;
-            return default;
-        }
-
-        IObjectReference NativeObject { get; }
-        bool HasUnwrappableNativeObject { get; }
-
-        protected ConcurrentDictionary<RuntimeTypeHandle, IObjectReference> QueryInterfaceCache { get; }
-
-        IObjectReference GetObjectReferenceForType(RuntimeTypeHandle type)
-        {
-            return GetObjectReferenceForTypeFallback(type);
-        }
-
-        IObjectReference GetObjectReferenceForTypeFallback(RuntimeTypeHandle type)
-        {
-            if (IsInterfaceImplemented(type, true))
-            {
-                return QueryInterfaceCache[type];
-            }
-            throw new Exception("Interface " + Type.GetTypeFromHandle(type) +" is not implemented.");
-        }
-
-        ConcurrentDictionary<RuntimeTypeHandle, object> AdditionalTypeData { get; }
-
-        object GetOrCreateTypeHelperData(RuntimeTypeHandle type, Func<object> helperDataFactory)
-        {
-            return AdditionalTypeData.GetOrAdd(type, (type) => helperDataFactory());
-        }
-    }
-}
\ No newline at end of file
+#pragma warning restore IL3050
+                if (!QueryInterfaceCache.TryAdd(interfaceType, typedObjRef))
+                {
+                    typedObjRef.Dispose();
+                }
+                return true;
+            }
+        }
+
+#if NET8_0_OR_GREATER
+        unsafe bool LookupGeneratedVTableInfo(RuntimeTypeHandle interfaceType, [NotNullWhen(true)] out IIUnknownCacheStrategy.TableInfo? result, out int qiResult)
+        {
+            result = null;
+            qiResult = 0;
+            if (AdditionalTypeData.TryGetValue(interfaceType, out object value))
+            {
+                if (value is IIUnknownCacheStrategy.TableInfo tableInfo)
+                {
+                    result = tableInfo;
+                    return true;
+                }
+                return false;
+            }
+
+            if (StrategyBasedComWrappers.DefaultIUnknownInterfaceDetailsStrategy.GetIUnknownDerivedDetails(interfaceType) is IIUnknownDerivedDetails details)
+            {
+                qiResult = NativeObject.TryAs(details.Iid, out ObjectReference<IUnknownVftbl> objRef);

```

Comment:
I was taking one last look at this before signing off and I noticed there is nothing that keeps the `objref` alive.  Do we want to add it to `QueryInterfaceCache` so that it keeps it alive?  The TableInfo just seems to point to `ThisPtr` which is an `IntPtr`.